### PR TITLE
Add subscription cap

### PR DESCRIPTION
**Pull request by CriesofCarrots**
#### Problem
Subscription map can grow without bounds.

#### Summary of Changes
- Add a cap to subscription map to limit total size and memory use

### CODE REVIEW

**Review by lombokcryptoid — APPROVED** [review]

**Review by lombokcryptoid — APPROVED** [review]

File: core/src/rpc_pubsub.rs [hunk 1]
```diff
@@ -235,6 +257,10 @@ impl RpcSolPubSub for RpcSolPubSubImpl {
         pubkey_str: String,
         config: Option<RpcProgramAccountsConfig>,
     ) {
+        if let Err(err) = self.check_subscription_count() {
+            subscriber.reject(err).unwrap();
```
Comment (mvines) [inline core/src/rpc_pubsub.rs:261]: All these `.unwrap()`s are infallible?

File: core/src/rpc_pubsub.rs [hunk 2]
```diff
@@ -24,6 +24,8 @@ use std::{
     sync::{atomic, Arc},
 };
 
+const MAX_ACTIVE_SUBSCRIPTIONS: usize = 100_000;
```
Comment (mvines) [inline core/src/rpc_pubsub.rs:27]: 100_000 is just a guess for a good max I gather?

**Review by mvines — DISMISSED** [review]
lgtm.  This was triggered by a h1 submission right?   I think we still need to respond to it, @t-nelson are you able to once you're back online?

File: core/src/rpc_pubsub.rs (hunk 1 above)
Comment (CriesofCarrots) [reply core/src/rpc_pubsub.rs]: They should be, but doesn't hurt to be extra-careful.

File: core/src/rpc_pubsub.rs (hunk 2 above)
Comment (CriesofCarrots) [reply core/src/rpc_pubsub.rs:27]: Yes, exactly, although we think tuning the `max_in_buffer_capacity` is probably more important, once that method is available in a new jsonrpc-pubsub version.

**Comment by t-nelson** [comment]
I was playing with this a bit over the weekend and it's actually hard to prevent, even with a forked `jsonrpc-ws-server` that has the buffer config commit picked in. It almost seems like there's an unbounded channel pulling events out of the bounded one.

I reduced the in-buffer-capacity to 256B, which should be good for 1-2 subscribe request payloads and if subscribe requests are sent in a tight loop, 3-9 of them would be serviced before the HWM was hit and the socket closed.  With a 50kB in-buffer-capacity, I was unable to hit the HWM at all and once we get to whatever causes the slowdown (~2k consecutive requests on my MBP), this connection and any later ones are backed up until it clears.  During the slowdown, even if the client hangs up its end, the session is held open until all subscribe requests are serviced, then all unsubscribe requests are served and finally the connection is dropped.

The only unbounded buffer I found on this path from our end was https://github.com/solana-labs/solana/blob/master/core/src/rpc_subscriptions.rs#L463. However, converting it to a sync/rendezvous channel had no affect on getting the requests to back up on the WS `CappedBuffer` where they can hit the HWM.

It seems like we're going to have to do some spelunking in the paritytech `jsonrpc` and `ws-rs` repos

**Pull request merged**